Add UML generation, cleanup + add documentation

MikeSmithEU opened the pull request:
**Why?**

Greatly improves upon the following guiding principles going forward:
 - Annotation over documentation. Documentation should relate to code where possible.
 - Pragmatism. We want to deliver maximum benefit with minimal effort, incrementally.

(Ref: https://github.com/ebu/benchmarkstt/wiki/Principles)


**Changes and additions**

- [x] Makefile: use environment python if available, otherwise use python3
- [x] Makefile: ensure pip is installed (in some cases needed for development, avoids user confusion)
- [x] cleanup code
- [x] add automatic UML diagrams to documentation
- [x] custom autodoc templates
- [x] add sphinx mermaid autoclassmemberdiagram directive
- [x] refactor code:
  - [x] group `cli` and `api` entrypoints in their respective packages
  - [x] use more descriptive names for `Base` classes (`Normalizer`, `Differ`, etc.)


**Chronologic todos**

- [x] Makefile: use environment python if available, otherwise use python3
- [x] Makefile: ensure pip is installed (in some cases needed for development, avoids user confusion)
- [x] removed an unused import from `src/benchmarkstt/cli.py`
- [x] adds additional documentation (per-package docstrings)
- [x] converts benchmarkstt.benchmark from a module to a namespace
- [x] make the magic per-package `api.py` and `cli.py` "private" (moved to `_api.py` and `_cli.py`)
- [x] autogenerate PlantUML files per package
- [x] add (simplified) custom PlantUML class diagrams per package to docs
- [x] implement a PlantUML renderer to SVG
- [x] link all packages and classes to proper documentation pages
- [x] optimize PlantUML svg renderer
- [x] omit _uml.py from code coverage checks
- [x] add uml generation to Makefile
- [x] generate uml on docs generation
- [x] document `plantuml.jar` requirement for generating uml graphs
  - [x] add in comments in `_uml.py`
  - [x] add explanation about missing `plantuml.jar` and where to download if plantuml call fails
- [x] `docs/CONTRIBUTING.rst`
  - [x] emphasize not pushing straight to master branch
  - [x] add info about license of code added by contributors 
- [x] Remove the plantuml.jar dependency (use  http://www.gravizo.com/ )
- [x] Support `.. plantuml:` or `.. gravizo:` ReStructuredText in documentation (https://github.com/MikeSmithEU/sphinxcontrib-gravizo)
- [x] cleanup all plantuml -> svg code 
- [x] update docs to use `.. gravizo:`
- [x] find work-around for too long request uri's for the bigger UML diagrams
- [x] add custom autodoc templates
- [x] update sphinx to latest
- [x] automate mermaid structure graphs
- [x] add members to autodocgraph (custom sphinx directive based on existing mermaid)
- [x] completely move to mermaid
- [x] cleanup all gravizo
- [x] cleanup `docs/conf.py`
- [x] generate svgs on server side for documentation
- [x] factor out yarn usage, simplify docs run (using `pynpm`)
- [x] add changelog

MikeSmithEU committed:
Fix makefile: use environment python if available, otherwise use python3
Add plantuml uml generation

MikeSmithEU committed:
add more readible umls

MikeSmithEU committed:
cleaner

MikeSmithEU committed:
cleanup

MikeSmithEU committed:
cleanup + add docs requirements to pip install

MikeSmithEU committed:
one deeper level

MikeSmithEU committed:
generate cleaner uml

MikeSmithEU committed:
dont cover uml.py

MikeSmithEU committed:
local rendering of plantuml files

MikeSmithEU committed:
lint

MikeSmithEU committed:
links (concept)

MikeSmithEU committed:
cli,api -> _cli,_api (make "protected", separates api and cli functionality from main class code)

MikeSmithEU committed:
cleanup

MikeSmithEU committed:
cleanup

MikeSmithEU committed:
cleanup, fixes and logging

MikeSmithEU committed:
more cleanup/fixes

MikeSmithEU committed:
more cleanup

MikeSmithEU committed:
regenerate

MikeSmithEU committed:
more cleanup

MikeSmithEU committed:
cleanup

MikeSmithEU committed:
better filtering of klass

MikeSmithEU committed:
cleaner formatting

MikeSmithEU committed:
make logger singleton, add support for namedtuples in plantuml, plantuml extension -> puml

MikeSmithEU committed:
add missing uml

MikeSmithEU committed:
hide methods that are inherited from outside package

MikeSmithEU committed:
proper documentation with uml, optimize uml, fix cleaning of modules rst autodocs

MikeSmithEU commented:
Generated docs available at https://benchmarkstt.readthedocs.io/en/uml-schema (hidden)

MikeSmithEU committed:
set a version specific to this branch

MikeSmithEU committed:
proper links for packages as well

MikeSmithEU committed:
cleanup/fixes

MikeSmithEU committed:
add explanation about missing plantuml.jar and where to download if plantuml call fails

MikeSmithEU committed:
Makefile fixes

MikeSmithEU committed:
cleanup begets cleanup

MikeSmithEU committed:
Avoid drawing packages that don't have any classes, cleanup some commented out old code

MikeSmithEU committed:
filter as param

MikeSmithEU committed:
more cleanup

MikeSmithEU committed:
annotate some magic methods as ducks

MikeSmithEU committed:
consistency

MikeSmithEU committed:
fix docs

MikeSmithEU committed:
bump version to 1.1.0.pre

MikeSmithEU committed:
add 2020 to copyright years

MikeSmithEU committed:
some extra ducks, sort ducks, regenerate uml

MikeSmithEU committed:
Make docs Makefile use env python if available or python3 if not. Fix documentation warning.

danielthepope commented on Makefile:
```diff
@@ -1,30 +1,47 @@
-testcoverage:
-	PYTHONPATH="./src/" python -m pytest --cov=./src tests/
+env:
+	test -e env/bin/activate && source env/bin/activate
```
This will fail if you don't have a virtual environment set up. Do you still need this line?

MikeSmithEU replied:
Good point, I've indeed replaced this with the more usable 
```PYTHON:=$(shell test -e env/bin/activate && echo "env/bin/python" || echo "python3")```

Nice catch, fixed.

danielthepope commented on Makefile:
```diff
@@ -1,30 +1,47 @@
+	$(PYTHON) -m pip install -e '.[test]'
+
+uml: env cleanuml
+	PYTHONPATH="./src/" $(PYTHON) src/benchmarkstt/_uml.py
```
You need to set the `PLANTUML_COMMAND` environment variable here for _uml.py to work

MikeSmithEU replied:
This will get resolved as I move the code away from that dependency (see https://github.com/ebu/benchmarkstt/pull/134#issuecomment-675614746 )

danielthepope commented on setup.py:
```diff
@@ -11,6 +11,10 @@ def from_file(filename):
+def filter_requirements(line):
+    return not line.startswith('-e')
+
+docs_require = list(filter(filter_requirements, from_file('docs/requirements.txt').split('\n')))
```
This is really neat! I've got some projects that use both a requirements.txt and a setup.py, and this seems a great way to keep the requirements in sync :)

MikeSmithEU replied:
This project really helped me learn how to write more maintainable `setup.py`, it can be a bit of a hassle sometimes, as I'm sure you know.

danielthepope commented on src/benchmarkstt/modules.py:
```diff
@@ -9,8 +9,8 @@
-    def __init__(self, sub_module):
-        self._postfix = '' if sub_module is None else '.' + sub_module
+    def __init__(self, sub_module=None):
+        self._postfix = '' if sub_module is None else '._' + sub_module
```
Would `Modules` ever be used by the public? It seems a bit odd having this extra `_` because the calls to, for example, `Modules('cli')` are no longer clear that it's actually looking in the `_cli` module. If we expect the public to be able to call `Modules` themselves, I'd prefer all current calls to name their modules more explicitly, i.e. `Modules('_cli')`

MikeSmithEU replied:
The point is to separate the core components from the functional command line interface, api and other unrelated modules such as those solely related to documentation generation.

I consider the core components to be the building blocks for a benchmark, i.e. Normalization, Metrics, etc.

That being said, this separation should indeed be clearer, for now this is the pragmatic solution in order to not overload the documentation and UMLs with mostly unrelated functionality that will never be user-facing (i.e. package developers only). 

I'll have a deeper think about this. Suggestions welcome!

MikeSmithEU replied:
All code related to documentation generation should be removed from the core package, which I've now done. There is only `_cli.py` and `_api.py`:

```
src/benchmarkstt/benchmark/_api.py
src/benchmarkstt/benchmark/_cli.py
src/benchmarkstt/metrics/_api.py
src/benchmarkstt/metrics/_cli.py
src/benchmarkstt/normalization/_api.py
src/benchmarkstt/normalization/_cli.py
src/benchmarkstt/api/_cli.py
```

Is this acceptable to you @danielthepope ?

MikeSmithEU replied:
Refactored it out...
The whole using the underscore to separate it was silly and just pointed to the needed refactor/cleanup.

danielthepope reviewed (dismissed):
Hey Mike! Thanks for these changes, the diagrams are great! It's good to have you back :)

I've added a few comments, and a couple of extras here:

- I didn't have PlantUML installed, so I think installation instructions should be added to development.rst, under "Building the Documentation". You could mention it's available via homebrew (`brew install plantuml`) or via https://plantuml.com/download.
- Since the UML files are automatically generated, I wonder whether they should be included in git? I think I'd prefer to see `docs/_static/uml` in the .gitignore instead (and create the `uml` directory in the Makefile)

MikeSmithEU commented:
> Hey Mike! Thanks for these changes, the diagrams are great! It's good to have you back :)
> 
> I've added a few comments, and a couple of extras here:
> 
> * I didn't have PlantUML installed, so I think installation instructions should be added to development.rst, under "Building the Documentation". You could mention it's available via homebrew (`brew install plantuml`) or via https://plantuml.com/download.
> * Since the UML files are automatically generated, I wonder whether they should be included in git? I think I'd prefer to see `docs/_static/uml` in the .gitignore instead (and create the `uml` directory in the Makefile)

It is my intention to remove the plantuml -> svg from this PR. I agree that the `uml` directory should be `.gitignore`d. I am currently having issues generating the svg's on readthedocs.org though, and have spent quite a lot of time looking for alternate solutions. The only reasonable one I've found thus far is http://www.gravizo.com/

Simply documenting how to install plantuml is not feasibly imho (or at least, I won't do it), as it would need to cover Windows, OSX and Linux targets, making sure a proper version of java is installed and provide ongoing support for this.

I've now converted this PR to draft until I can implement that, but it gets me to quite a side tangent as this is far outside the scope of this project, and the cleanest is writing a new sphinx plugin for this. I am really short on free time at the moment so this is appended to my long todo list. To be continued...

MikeSmithEU committed:
feedback: https://github.com/ebu/benchmarkstt/pull/134\#discussion_r469075601

MikeSmithEU committed:
sphinxcontrib-gravizo

MikeSmithEU committed:
bye bye plantuml.jar dependency

MikeSmithEU commented:
@danielthepope  Embed PlantUML, DOT, etc. diagrams in documentation using Gravizo now supported, and separated it from benchmarkstt.

https://pypi.org/project/sphinxcontrib-gravizo/

Examples
Inline graph, show as png:

```
.. gravizo:: png
    @startuml
    Alice -> Bob: Authentication Request
    Bob --> Alice: Authentication Response

    Alice -> Bob: Another authentication Request
    Alice <-- Bob: Another authentication Response
    @enduml
```

Load from a file, show as svg:

```
.. gravizo:: ./path/to/graph.puml svg
```

MikeSmithEU committed:
cleanup

MikeSmithEU committed:
cleanup

MikeSmithEU committed:
use gravizo

MikeSmithEU committed:
big cleanup

MikeSmithEU committed:
use new gravizo version

MikeSmithEU commented:
Docs build of uml files is now completely separated. Gravizo plugin moved to sphinxcontrib organization and added as a dependency for docs. All svg- and puml-files are no longer polluting this repository.

Since there are a few big graphs that cannot be directly generated by a GET-request to Gravizo (request uri too long as per HTTP RFC) I still need to implement a work-around for that.

MikeSmithEU committed:
.coveragerc omit no longer required

MikeSmithEU committed:
import after sys path set

MikeSmithEU committed:
fix

MikeSmithEU committed:
Squashed commit of the following:

commit 9549761be5a04417fc160744f6d34dde325838c8
Author: MikeSmithEU <github@mikesmith.eu>
Date:   Fri Aug 21 01:41:34 2020 +0000

    unneeded afaik

commit 1a98f684a384bf4c0f429fec4a91dab358314885
Author: MikeSmithEU <github@mikesmith.eu>
Date:   Fri Aug 21 01:07:36 2020 +0000

    readable structures

commit e7b3143d031fa1421a60b2a96a0c4b4d9281f53d
Author: MikeSmithEU <github@mikesmith.eu>
Date:   Thu Aug 20 22:28:49 2020 +0000

    templates

commit 0e3fa8c3fa38d878caa8d84d54243d46ae8476b7
Author: MikeSmithEU <github@mikesmith.eu>
Date:   Thu Aug 20 22:26:44 2020 +0000

    tmp

MikeSmithEU committed:
refactor

MikeSmithEU committed:
more refactor, reflect in docs

MikeSmithEU committed:
refactor, deferred

MikeSmithEU committed:
more cleanup, improve docs

MikeSmithEU committed:
test

MikeSmithEU committed:
more cleanup

MikeSmithEU committed:
shrug

MikeSmithEU committed:
cleanup for github

MikeSmithEU committed:
fix

MikeSmithEU committed:
fix

MikeSmithEU committed:
ignore notebooks

MikeSmithEU committed:
cleaner

MikeSmithEU committed:
mermaid with class members

MikeSmithEU committed:
add associations

MikeSmithEU committed:
better per-package Base class names

MikeSmithEU committed:
cleanup

MikeSmithEU committed:
cleanup

MikeSmithEU committed:
lint

MikeSmithEU committed:
no more gravizo usage

MikeSmithEU committed:
cleanup

MikeSmithEU committed:
uml.py no longer needed, migrated all to mermaid

MikeSmithEU committed:
delete code ftw

danielthepope commented on Makefile:
```diff
@@ -1,30 +1,40 @@
+
+setupdocs: env
+	$(PYTHON) -m pip install -r docs/requirements.txt
+	cd docs/ && yarn install
```
Is there much of a benefit to using yarn over npm? It's just an extra dependency that I need to install manually, and I don't think it's needed since you've got a package-lock.json there already.

MikeSmithEU replied:
Mostly yarn is the preferred package manager for many nodejs packages. The reason I used it here is because it is what mermaid-js itself documents in its install instructions. Since it is only for documentation generation I do not see the extra dependency on yarn as a big issue. 
As far as benefits, there are claims of improved speed, security and stability that I haven't gone deeper into. But truth be told, since it is not part of the main benchmarkstt code, I have no preference either way. The claims of yarn dependency on the mermaid-js installation instructions makes me more inclined to keep using yarn though.

Ref: https://mermaid-js.github.io/mermaid/#/?id=installation

MikeSmithEU replied:
(removed the `package-lock.json` as it was a left-over of installing `mermaid-cli` during my first attempt to use mermaid without using the sphinxcontrib-mermaid extension)

MikeSmithEU replied:
(removed `yarn install` from this Makefile as it is already being run in the docs/Makefile)

MikeSmithEU replied:
I've now factored out `yarn` completely and used `pynpm` to do the `mermaid` installation.

danielthepope reviewed (commented):
That's very cool how you've managed to make the diagrams generate in the front end! I wonder if I'm doing something wrong though, or maybe I've spotted a problem?

After generating the docs, running `cd docs/build/html; python3 -m http.server` I go to http://localhost:8000/modules/benchmarkstt.api.html.

A diagram doesn't show up and there's an error in the browser's console
```
Uncaught Error: Parse error on line 1:
classDiagram
------------^
Expecting 'NEWLINE', got 'EOF'
```

MikeSmithEU commented:
> That's very cool how you've managed to make the diagrams generate in the front end! I wonder if I'm doing something wrong though, or maybe I've spotted a problem?
> 
> After generating the docs, running `cd `docs/build/html;` python3 -m http.server` I go to http://localhost:8000/modules/benchmarkstt.api.html.
> 
> A diagram doesn't show up and there's an error in the browser's console
> 
> ```
> Uncaught Error: Parse error on line 1:
> classDiagram
> ------------^
> Expecting 'NEWLINE', got 'EOF'
> ```

Sorry for the late reply, I've been indisposed for a couple of days.

Indeed, this is because for those modules there are no public classes available, so it returns an "empty" classDiagram, i.e. an unexpected EOF. I will have a look if I can either pre-check this (probably will result in messy code fixes), or contribute the functionality to "skip if empty" to the original sphinxcontrib repository (seems the better idea, but does not assure acceptance by the repo maintainer(s)). I feel this can be ignored for the time being as it does not really affect the users of the documentation.

MikeSmithEU committed:
minor cleanup

MikeSmithEU committed:
Remove `yarn install` from /Makefile::setupdocs as it is already included in the docs/Makefile

danielthepope reviewed (commented):
> I feel this can be ignored for the time being as it does not really affect the users of the documentation.

Fair enough, that's fine then.

I've just realised that because you're using this new way of generating UML, you no longer need the `autogen` folder, so you can delete the `docs/_static/autogen/.keep` file.

After that, this will be good to merge 😄

MikeSmithEU committed:
remove no longer needed docs/_static/autogen/.keep

MikeSmithEU commented:
> > I feel this can be ignored for the time being as it does not really affect the users of the documentation.
> 
> Fair enough, that's fine then.
> 
> I've just realised that because you're using this new way of generating UML, you no longer need the `autogen` folder, so you can delete the `docs/_static/autogen/.keep` file.
> 
> After that, this will be good to merge 😄

Good catch, I've been experimenting with generating the svg's server side though, and thought it might still be useful (turns out even then it's not needed). Cleaned them up in the latest commit.

~~The svg generating server side is quite problematic it seems. It spews all kinds of weird syntax errors on the mermaid code (which does render properly using the client-side renderer or mermaid live preview). Some are related to html escaping (which I solved locally), others I haven't been able to figure out yet. Because I spent way too much time on it already I will keep it as-is.~~ (edit: turns out I was using the wrong package, `mermaid.cli` != `@mermaid-js/mermaid-cli`)

~~If we can agree on the usage of yarn, I suggest this to be the ready-to-merge PR, pending approval from you.~~ edit: Factored out `yarn`

~~Edit: let's consider this a near-ready-to-merge PR, there are a few issues I still need to solve. I.e. the textwrapping of the `__init__` arguments in the diagram shows some issues that I need to fix. And there is some self-referencing arrows in the diagrams that shouldn't be there. Both these issues should be quite easy to fix (requiring much less time and effort than any of the experiments I've done before to reach this point).~~
~~When these two remaining issues are taken care of (by the end of the week for sure),~~ (edit: fixed) I will merge once you approve of my changes.

MikeSmithEU committed:
remove unnecessary clearing of removed directory

MikeSmithEU committed:
final fixes

MikeSmithEU committed:
final final fixes

MikeSmithEU commented:
@danielthepope This should be ready to merge once you approve...

MikeSmithEU committed:
remove unnessary import

MikeSmithEU committed:
test for readthedocs

MikeSmithEU committed:
rtfd

MikeSmithEU committed:
rtfd

MikeSmithEU committed:
use pynpm ¯\_(ツ)_/¯

MikeSmithEU committed:
optimize

MikeSmithEU commented:
Factored out `yarn`

MikeSmithEU committed:
test rtfd with package-lock and nosandbox

MikeSmithEU committed:
try without parallel install of npm packages

MikeSmithEU committed:
test

MikeSmithEU committed:
debug output for mmdc, use absolute paths

MikeSmithEU committed:
output node version

MikeSmithEU committed:
update node from python?

MikeSmithEU committed:
sudo

MikeSmithEU committed:
might as well use node v14

MikeSmithEU committed:
try latest docker

MikeSmithEU committed:
no local builds for rtfd :(

MikeSmithEU committed:
mermaid 8.7.0

MikeSmithEU committed:
dotted line for composition

MikeSmithEU committed:
add some type hints

MikeSmithEU committed:
cleanup

MikeSmithEU committed:
add changelog

MikeSmithEU committed:
make pynpm optional, render uml locally by default

MikeSmithEU committed:
changelog to rst

MikeSmithEU committed:
add changelog to docs, fix typo
